False positive `inert_defer` with condition compilation

### New Issue Checklist

- [x] Updated SwiftLint to the latest version
- [x] I searched for [existing GitHub issues](https://github.com/realm/SwiftLint/issues)

### Describe the bug

Rule `inert_defer` is false positive when defer is wrapped in condition compilation. For example: `#if os(macOS)`

##### Complete output when running SwiftLint, including the stack trace and command used

```bash
$ swiftlint Test.swift

Linting Swift files at paths Test.swift
Linting 'Test.swift' (1/1)
.../Test.swift:6:9: warning: Inert Defer Violation: If defer is at the end of its parent scope, it will be executed right where it is anyway. (inert_defer)
Done linting! Found 1 violation, 0 serious in 1 file.
```

### Environment

* SwiftLint version - `0.50.1`
* Installation method used  - Homebrew
* Paste your configuration file - default 


* Xcode version - `Xcode 14.0.1 Build version 14A400`
* Run - `swiftlint lint --path Test.swift --no-cache --enable-all-rules`
```
../Test.swift:4:5: warning: Explicit ACL Violation: All declarations should specify Access Control Level keywords explicitly. (explicit_acl)
../Test.swift:6:9: warning: Inert Defer Violation: If defer is at the end of its parent scope, it will be executed right where it is anyway. (inert_defer)
../Test.swift:3:8: warning: Missing Docs Violation: public declarations should be documented. (missing_docs)
../Test.swift:3:8: warning: Required Deinit Violation: Classes should have an explicit deinit method. (required_deinit)
```

```swift
public class Test {
    func hello(name: String) -> String {
        #if os(macOS)
        // this trigger violation: Inert Defer Violation: If defer is at the end of its parent scope, it will be executed right where it is anyway. (inert_defer)
        defer {
            print(self)
        }
        #endif

        print(name)

        return "Hello " + name
    }
}
```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -43,6 +43,11 @@
   [SimplyDanny](https://github.com/SimplyDanny)
   [#4612](https://github.com/realm/SwiftLint/issues/4612)
 
+* Skip `defer` statements being last in an `#if` block if the `#if` statement is
+  not itself the last statement in a block.  
+  [SimplyDanny](https://github.com/SimplyDanny)
+  [#4615](https://github.com/realm/SwiftLint/issues/4615)
+
 * Fix false positives in `empty_enum_arguments` when the called expression
   is an identifier or an init call.  
   [Steffen Matthischke](https://github.com/heeaad)

--- a/Source/SwiftLintFramework/Rules/Lint/InertDeferRule.swift
+++ b/Source/SwiftLintFramework/Rules/Lint/InertDeferRule.swift
@@ -25,7 +25,17 @@ struct InertDeferRule: ConfigurationProviderRule, SwiftSyntaxRule {
                     print("other code")
                 }
             }
-            """)
+            """),
+            Example("""
+            func f() {
+                #if os(macOS)
+                defer { print(2) }
+                #else
+                defer { print(3) }
+                #endif
+                print(1)
+            }
+            """, excludeFromDocumentation: true)
         ],
         triggeringExamples: [
             Example("""
@@ -46,7 +56,23 @@ struct InertDeferRule: ConfigurationProviderRule, SwiftSyntaxRule {
                     // comment
                 }
             }
-            """)
+            """),
+            Example("""
+            func f(arg: Int) {
+                if arg == 1 {
+                    ↓defer { print(2) }
+                    // a comment
+                } else {
+                    ↓defer { print(3) }
+                }
+                print(1)
+                #if os(macOS)
+                ↓defer { print(4) }
+                #else
+                ↓defer { print(5) }
+                #endif
+            }
+            """, excludeFromDocumentation: true)
         ]
     )
 
@@ -58,13 +84,25 @@ struct InertDeferRule: ConfigurationProviderRule, SwiftSyntaxRule {
 private extension InertDeferRule {
     final class Visitor: ViolationsSyntaxVisitor {
         override func visitPost(_ node: DeferStmtSyntax) {
-            guard let codeBlockItem = node.parent?.as(CodeBlockItemSyntax.self),
-                  let codeBlockList = codeBlockItem.parent?.as(CodeBlockItemListSyntax.self),
-                  codeBlockList.last == codeBlockItem  else {
+            guard node.isLastStatement else {
+                return
+            }
+            if let ifConfigClause = node.parent?.parent?.parent?.as(IfConfigClauseSyntax.self),
+               ifConfigClause.parent?.parent?.isLastStatement == false {
                 return
             }
 
             violations.append(node.deferKeyword.positionAfterSkippingLeadingTrivia)
         }
+    }
+}
+
+private extension SyntaxProtocol {
+    var isLastStatement: Bool {
+        if let codeBlockItem = parent?.as(CodeBlockItemSyntax.self),
+           let codeBlockList = codeBlockItem.parent?.as(CodeBlockItemListSyntax.self) {
+            return codeBlockList.last == codeBlockItem
+        }
+        return false
     }
 }

